### PR TITLE
Cloning replica to use async zookeeper calls

### PR DESCRIPTION
**Pull request by canhld94**
This can help to reduce startup time when tables need to be cloned from other replicas.

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Use async ZooKeeper APIs during cloning replica to reduce cloning replica time.

When a replica lag is too high and normal replication via log queue cannot work, `StorageReplicatedMergeTree` will try to clone other replica by: (1) get all active parts from a source replica, (2) create log entries for parts that are in source replica but not current replica, and (3) copy the log entries from source replica log queue to current replica log queue.

Both (1) (2) and (3) makes a lots of requests to zookeeper, and they're now using synchronous calls within a tight for-loop. Therefore, when zookeeper latency is high and there're many missing parts, cloneReplica may take several hours. If cloning replica happens when server restart, it will block the server from starting up.

This PR reduces `StorageReplicatedMergeTree::cloneReplica` execution time by using async zookeeper calls.

This issue did happen to us last week, a server was down for half a day and it took two hours to start up. We dumped all thread stacktrace during start-up and most of them stuck in `cloneReplica`:

```
Thread 34 (Thread 0x7f2a863f9700 (LWP 2766066) "BgSchPool"):
#0  futex_abstimed_wait_cancelable (private=0, abstime=0x7f2a863efdc8, clockid=-2042692272, expected=0, futex_word=0x7ecf68e6d168) at ../sysdeps/nptl/futex-internal.h:323
#1  __pthread_cond_wait_common (abstime=0x7f2a863efdc8, clockid=-2042692272, mutex=0x7ecf68e6d118, cond=0x7ecf68e6d140) at pthread_cond_wait.c:520
#2  __pthread_cond_timedwait (cond=0x7ecf68e6d140, mutex=0x7ecf68e6d118, abstime=0x7f2a863efdc8) at pthread_cond_wait.c:656
#3  0x0000000010e88206 in std::__1::__libcpp_condvar_timedwait[abi:v15000](pthread_cond_t*, pthread_mutex_t*, timespec*) (__cv=0x7ecf68e6d140, __m=0x7ecf68e6d118, __ts=0x7f2a863efdc8) at ./contrib/llvm-project/libcxx/include/__threading_support:341
#4  std::__1::condition_variable::__do_timed_wait (this=0x7ecf68e6d140, tp=..., lk=...) at ./contrib/llvm-project/libcxx/src/condition_variable.cpp:76
#5  0x0000000010e88206 in std::__1::condition_variable::wait_for<long long, std::__1::ratio<1l, 1000000000l> > (this=0x7ecf68e6d140, __lk=..., __d=...)
#6  std::__1::condition_variable::__do_timed_wait<std::__1::chrono::steady_clock> (this=0x7ecf68e6d140, __tp=..., __lk=...) at ./contrib/llvm-project/libcxx/include/__mutex_base:512
#7  std::__1::condition_variable::wait_until<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l> > > (this=0x7ecf68e6d140, __t=..., __lk=...) at ./contrib/llvm-project/libcxx/include/__mutex_base:415
#8  std::__1::__assoc_sub_state::wait_until<std::__1::chrono::steady_clock, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l> > > (this=0x7ecf68e6d100, __abs_time=...) at ./contrib/llvm-project/libcxx/include/future:611
#9  0x0000000012803459 in std::__1::__assoc_sub_state::wait_for[abi:v15000]<long long, std::__1::ratio<1l, 1000l> >(std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000l> > const&) const (this=0x7ecf68e6d100, __rel_time=...) at ./contrib/llvm-project/libcxx/include/future:622
#10 std::__1::future<Coordination::GetResponse>::wait_for[abi:v15000]<long long, std::__1::ratio<1l, 1000l> >(std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000l> > const&) const (this=<optimized out>, __rel_time=...) at ./contrib/llvm-project/libcxx/include/future:1092
#11 zkutil::ZooKeeper::getImpl(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, Coordination::Stat*, std::__1::function<void (Coordination::WatchResponse const&)>) (this=0x7f2989bdf598, path=..., res=..., stat=0x7f2a863f01d0, watch_callback=...) at ./build/./src/Common/ZooKeeper/ZooKeeper.cpp:524
#12 0x0000000012803adf in zkutil::ZooKeeper::tryGetWatch(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, Coordination::Stat*, std::__1::function<void (Coordination::WatchResponse const&)>, Coordination::Error*) (this=0x7ecf68e6d168, path=..., res=..., stat=0x7f2a863f01d0, watch_callback=..., return_code=0x0) at ./build/./src/Common/ZooKeeper/ZooKeeper.cpp:580
#13 0x00000000128039c8 in zkutil::ZooKeeper::tryGet (this=0x7f2989bdf598, path=..., res=..., stat=0x7f2a863f01d0, watch=..., return_code=<optimized out>) at ./build/./src/Common/ZooKeeper/ZooKeeper.cpp:570
#14 0x00000000116f2b52 in DB::tryGetPartCreateTime (replica_path=..., part_name=..., zookeeper=...) at ./build/./src/Storages/StorageReplicatedMergeTree.cpp:1402
#15 DB::StorageReplicatedMergeTree::cloneReplica (this=0x7ecf4fea7040, source_replica=..., source_is_lost_stat=..., zookeeper=...) at ./build/./src/Storages/StorageReplicatedMergeTree.cpp:3226
#16 0x00000000116fdca1 in DB::StorageReplicatedMergeTree::cloneReplicaIfNeeded (this=0x7ecf4fea7040, zookeeper=...) at ./build/./src/Storages/StorageReplicatedMergeTree.cpp:3468
#17 0x0000000011e80b16 in DB::ReplicatedMergeTreeRestartingThread::tryStartup (this=0x7ecf4fea8288) at ./build/./src/Storages/MergeTree/ReplicatedMergeTreeRestartingThread.cpp:168
#18 0x0000000011e7e950 in DB::ReplicatedMergeTreeRestartingThread::runImpl (this=0x7ecf4fea8288) at ./build/./src/Storages/MergeTree/ReplicatedMergeTreeRestartingThread.cpp:132
#19 DB::ReplicatedMergeTreeRestartingThread::run (this=0x7ecf4fea8288) at ./build/./src/Storages/MergeTree/ReplicatedMergeTreeRestartingThread.cpp:58
#20 0x00000000117208bf in DB::StorageReplicatedMergeTree::startupImpl (this=0x7ecf4fea7040, from_attach_thread=true) at ./build/./src/Storages/StorageReplicatedMergeTree.cpp:5141
#21 0x0000000011dfefde in DB::ReplicatedMergeTreeAttachThread::finalizeInitialization (this=0x7ecf4fea8300) at ./build/./src/Storages/MergeTree/ReplicatedMergeTreeAttachThread.cpp:205
#22 DB::ReplicatedMergeTreeAttachThread::run (this=0x7ecf4fea8300) at ./build/./src/Storages/MergeTree/ReplicatedMergeTreeAttachThread.cpp:56
#23 0x000000000fd8973f in std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const (this=0x7ecf68e6d168) at ./contrib/llvm-project/libcxx/include/__functional/function.h:848
```

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 495c7bb8c0deca526181a80c400d34dcc12650e0 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59324/495c7bb8c0deca526181a80c400d34dcc12650e0/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59324/495c7bb8c0deca526181a80c400d34dcc12650e0/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by canhld94** [comment]
Pinging for reviewing this PR @alexey-milovidov @tavplubix. It's very helpful when the latency to zookeeper is high (it's an unusual situation, but not that it never happen).
FTR, We have been running ClickHouse with this PR for several months. We have a cluster with relatively high (continent-to-continent) latency to zk. Sometime a machine will be taken down for one day for maintaining and replicas on that machine will be lost. It used to take two hours to start up ClickHouse, but now it's less than five minutes!

**Review by alexey-milovidov — APPROVED** [review]
The code looks ok to me. Let's merge with master.
Also, I'd appreciate another review from @tavplubix

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -3231,11 +3222,58 @@ void StorageReplicatedMergeTree::cloneReplica(const String & source_replica, Coo
...
+            }
+        }
+        LOG_TEST(log, "Enqueueing {} for fetch", entry->new_part_name);
+        futures.emplace_back(zookeeper->asyncCreate(fs::path(replica_path) / "queue/queue-", entry->toString(), zkutil::CreateMode::PersistentSequential));
+    }
+
+    for (const auto & index : copy_entries)
+    {
+        LOG_TEST(log, "Copying entry {}", source_queue[index].data);
+        futures.emplace_back(zookeeper->asyncCreate(fs::path(replica_path) / "queue/queue-", source_queue[index].data, zkutil::CreateMode::PersistentSequential));
```
Comment (tavplubix) [inline src/Storages/StorageReplicatedMergeTree.cpp:3256]: Consider batching create requests and sending them as a multi request (however, Apache ZooKeeper has 1MB limit for multi-request size, but sending MULTI_BATCH_SIZE requests at once should be fine)

**Comment by alexey-milovidov** [comment]
No activity since the review.

**Pull request closed without merging**